Port adaptive_max_pool2d() to ATen

skrah opened the pull request:
This is the first part of  #18064.

skrah committed:
Port adaptive_max_pool2d() to ATen.

skrah committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into adaptive_max_pooling

ezyang commented:
@xmnlab, could you do first pass review on this PR? I'll look at it after you're done. Thanks!

skrah commented:
@ezyang There are still sporadic failures in `test_jit`, which unfortunately only occur when running all of `test_jit` but not when running a single test.

So this is still WIP.

ezyang commented:
Might be a memory error. Keep an eye on the pytorch_linux_xenial_py3_clang5_asan_test it might fail with some more useful information.

skrah committed:
Add workaround for test_jit.

  - The JIT passes an IntArrayRef of size 1 if AdaptiveMaxPool2d() is called
    with a single integer output_size parameter.

skrah committed:
Fix argument position.

skrah commented:
The cause of the `test_jit` failure is resolved.  If a module is constructed with a single `output_size` integer argument, the JIT passes that on as an `IntArrayRef` *with size one*.  This is the specific test from `common_nn.py`:

```python
dict(
    module_name='AdaptiveMaxPool2d',
    constructor_args=(3,),
    input_fn=lambda: _rand_tensor_non_equal(1, 3, 5, 6),
    desc='single',
)
```

Since all code in `aten/src/ATen/native` expects valid `output_size[0]` and `output_size[1]`, the test crashed.

I'm not sure what the function contract is. The signature in `func: adaptive_max_pool2d(Tensor self, int[2] output_size) -> (Tensor, Tensor)` *does* specify an `output_size==2`.


I added a workaround that handles `output_size==1`, and the tests pass.

Note that `AdaptiveAvgPool2d` passes a similar test, but `adaptive_avg_pool2d` seems to be special cased in the JIT code, whereas `adaptive_max_pool2d` is not.

skrah commented:
I think this is ready for review. Note that I tried to follow the porting guide and optimized for a minimal diff size. So there are definitely passages that one might want to rewrite, but then the review gets much more complex.

xmnlab commented:
@ezyang sure I will review that today. Thanks

ezyang commented:
@xmnlab Thanks. Assign me as reviewer when you're done.

ezyang commented:
>  If a module is constructed with a single output_size integer argument, the JIT passes that on as an IntArrayRef with size one.

A few things:

* It's the responsibility for the kernel to do appropriate error checking; e.g., if we expect a size two array, this must be checked at some point. I do have trouble remembering where exactly this checking should occur; it's possible there was some automatic checking in the old TH code that's not in the new codepath.
 * When there is a specification with an explicit size, e.g., `int[2]`, this affects our Python bindings in the following way: "int[] accepts an optional length specifier, e.g., int[2], which has no effect in C++ but extends our Python bindings to accept a bare number, which will be expanded into an appropriately sized list by repeating the number" as per documentation in https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#func So it seems to me that there might be some JIT bug here.

> adaptive_avg_pool2d seems to be special cased in the JIT code, whereas adaptive_max_pool2d is not

Where is this special-casing, may I ask? I grepped for `adaptive_avg_pool2d` but I didn't see anything that looked like a special case.

skrah commented:
> [...] So it seems to me that there might be some JIT bug here.

Okay, thanks!


> Where is this special-casing, may I ask? I grepped for `adaptive_avg_pool2d` but I didn't see anything that looked like a special case.

I'm not sure if it has any effect on the issue at hand, but there's this in `torch/csrc/` for `adaptive_avg_pool2d` but not for `adaptive_max_pool2d`:

```

passes/shape_analysis.cpp:933:            "aten::adaptive_avg_pool2d(Tensor self, int[] output_size) -> Tensor",

symbolic_script.cpp:957:        def AD_adaptive_avg_pool2d_backward(grad,
symbolic_script.cpp:964:                grad_self = torch._adaptive_avg_pool2d_backward(grad, self)
symbolic_script.cpp:972:            grad_input = AD_adaptive_avg_pool2d_backward(grad.unsqueeze(2), input.unsqueeze(2), output_size_2d).squeeze(2)
symbolic_script.cpp:983:        def adaptive_avg_pool2d(self,
symbolic_script.cpp:987:                grad_self = AD_adaptive_avg_pool2d_backward(grad_output, self, output_size)
symbolic_script.cpp:989:            return torch.adaptive_avg_pool2d(self, output_size), backward
symbolic_script.cpp:1191:                grad_input = AD_adaptive_avg_pool2d_backward(grad, input, output_size)
```

xmnlab commented on aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:
```diff
@@ -0,0 +1,362 @@
+  const Tensor& input,
+  IntArrayRef output_size)
+{
+  Tensor output = at::empty({0}, input.options());
```
I think it also could use `auto` here, such as https://github.com/pytorch/pytorch/pull/14714/files#diff-257f62854d62776a19a5480585a39962R299 not sure what would be the best practice here.

skrah replied:
I don't know the style guide,  this one is from `FractionalMaxPool2d.cpp`.

ezyang replied:
I don't really mind, but I know @mike7ant prefers it when you don't use auto ;)

xmnlab commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+ *    this function adaptively maxpools an input 4D tensor along dimensions 2 and 3
+ *    4D input, 4D output, 4D argmax x and y
+ */
+ template <typename T>
```
using `T` here would be fine as it is also used in https://github.com/pytorch/pytorch/pull/14714/files#diff-7b64cf245203cfc8f04802bd74787113R36 .. but maybe would be good to change that to `scalar_t` as used in other files such as https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Embedding.cu#L29

ezyang replied:
Yeah, I don't think it's a big deal either way, people can figure it out.

xmnlab commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+      T *ptr_output = output + oh*osizeW + ow;
+      int64_t *ptr_ind = indices + oh*osizeW + ow;
+      int argmax = -1;
+      T max = THCNumerics<T>::min();
```
In the cuda files I found references to `THCNumerics` and also to `at::numeric_limits` ... not sure if here it could be changed to `at::numeric_limits<T>::min()`

ezyang replied:
We should... stop using THCNumerics. But there's a lot of use sites already in the code, so you don't have to fix it.

xmnlab commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+                                   istrideD, istrideH, istrideW);
+      } 
+    );
+    THCudaCheck(cudaGetLastError());
```
I think here you could change to `AT_CHECK(cudaGetLastError() == cudaSuccess, "Failed with error code ", cudaGetLastError())`

ezyang replied:
I think `AT_CUDA_CHECK(cudaGetLastError())` is the most idiomatic, but we are also horribly inconsistent in native today as well and so I think it's not a big deal if it stays this way too.

xmnlab commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+        }
+      }
+    );
+    THCudaCheck(cudaGetLastError());
```
same as commented above

xmnlab commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+        }
+      }
+    );
+    THCudaCheck(cudaGetLastError());
```
same as commented above

xmnlab reviewed (commented):
LGTM! just added some comments, thanks @skrah !
@ezyang could you check the comments I made? I am curious about the usage of `int` or `int64_t`

skrah commented on aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:
```diff
@@ -0,0 +1,362 @@
+namespace {
+
+inline int start_index(int a, int b, int c) {
+  return (int)std::floor((float)(a * c) / b);
```
That was my guess, too, but `AdaptiveAveragePooling.cpp` uses `int` and I did not want to enter a numerical analysis regarding the `float` cast. :)

ezyang replied:
Yeah, in this case, keeping it as `int` as before is fine. This might be something to look into later, but for now we should keep it as it did before.

skrah commented on aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:
```diff
@@ -0,0 +1,362 @@
+          const Tensor& input,
+          IntArrayRef output_size)
+{
+  int dimW = 2;
```
Well, `dimW` only takes the values 2 or 3, so `int` is fine.

ezyang replied:
I suppose there is a very nitpicky argument to be made about how it might be better to use `int64_t` because it prevents someone else from copypasting your code and then committing a "oops I used 32-bit indexing when I should have used 64-bit indexing error". I do think this is pretty benign, so I'll leave it up to you @skrah.

ezyang commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+#include <cfloat>
+#include <cmath>
+
+#define CUDA_MAX_THREADS 1024   // this is safe, in reality 256 is our limit
```
This define looks dead.

skrah replied:
Indeed, it should be removed.

ezyang commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+#include "ATen/Utils.h"
+#include "c10/util/Exception.h"
+#include <THC/THCGeneral.h>
+#include "THC/THCNumerics.cuh"
```
Could you make all of these includes used angled brackets?

ezyang commented on aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:
```diff
@@ -0,0 +1,472 @@
+#define CUDA_MAX_THREADS 1024   // this is safe, in reality 256 is our limit
+
+#define START_IND(a,b,c) (int)std::floor((float)(a * c) / b)
+#define END_IND(a,b,c) (int)std::ceil((float)((a + 1) * c) / b)
```
Macro here is suboptimal, but it doesn't have to get fixed in this PR.

ezyang replied:
Oh, but I do see you fixed it in your other file ;)

skrah replied:
@ezyang I tried it now for the follow-up PR, but CUDA does not like it (apparently the function is not inlined):

```
error: calling a __host__ function("at::native::_NV_ANON_NAMESPACE::start_index") from a __global__ function("at::native::_NV_ANON_NAMESPACE::adaptivemaxpool<double> ") is not allowed
```

skrah replied:
Sorry for the noise, obviously it needs `__device__`. :)

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3621,18 +3621,30 @@
 
 # Return: (Tensor output, Tensor indices)
-- func: adaptive_max_pool2d(Tensor self, int[2] output_size, *, Tensor(a!) output, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
+- func: adaptive_max_pool2d(Tensor self, int[2] output_size, *, Tensor(a!) out, Tensor(b!) indices) -> (Tensor(a!), Tensor(b!))
```
Huh, what happened here?

skrah replied:
That was me being in literal mode while trying to debug the `test_jit` issue. :) I was grasping at straws, then I saw that the README  recommends `out` as the convention, so I checked if the JIT by any chance relies on the argument name convention.

ezyang replied:
@skrah Nice catch. Could you add this to the wiki? https://github.com/pytorch/pytorch/wiki/TH-to-ATen-porting-guide You should have edit rights.

skrah replied:
@ezyang  Yes, I added a small entry to the wiki.

ezyang commented:
I don't see any substantive problems, only nits. This is the bees knees, thank you!

skrah commented:
Thank you!  Regarding the open comments, perhaps the easiest thing is to make a followup PRs.  For the CUDA macros and includes it will touch both `Avg` and `Max` pooling, for `THCNumerics` etc. it could be done for all files in `native` at once.

ezyang commented on aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:
```diff
@@ -0,0 +1,362 @@
+  #pragma omp parallel for private(b)
+    for (b = 0; b < sizeB; b++)
+    {
+      AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool2d_cpu", [&] {
```
Dynamic dispatch inside of an OMP parallelized loop!

ezyang commented on aten/src/ATen/native/AdaptiveMaxPooling2d.cpp:
```diff
@@ -0,0 +1,362 @@
+#pragma omp parallel for private(b)
+    for (b = 0; b < sizeB; b++)
+    {
+      AT_DISPATCH_FLOATING_TYPES(input.scalar_type(), "adaptive_max_pool2d_backward", [&] {
```
Dynamic dispatch inside an OMP parallelized loop!
